packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Models/Types/SystemModelSerializationTypeProvider.cs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ClientModel.Primitives;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+using Microsoft.Generator.CSharp.Expressions;
+
+namespace Microsoft.Generator.CSharp.ClientModel
+{
+    /// <summary>
+    /// This class provides the set of serialization models, methods, and interfaces for a given model.
+    /// </summary>
+    internal sealed class SystemModelSerializationTypeProvider : TypeProvider
+    {
+        internal static readonly Parameter SerializationOptionsParameter =
+            new("options", null, typeof(ModelReaderWriterOptions), null, ValidationType.None, null);
+
+        public SystemModelSerializationTypeProvider(ModelTypeProvider model) : base(null)
+        {
+            Model = model;
+            Name = model.Name;
+            Json = BuildJsonSerialization();
+            // Initialize the serialization interfaces
+            IJsonModelTInterface = new CSharpType(typeof(IJsonModel<>), Model.Type);
+            IJsonModelObjectInterface = Model.IsStruct ? new CSharpType(typeof(IJsonModel<object>)) : null;
+            IPersistableModelTInterface = new CSharpType(typeof(IPersistableModel<>), Model.Type);
+            IPersistableModelObjectInterface = Model.IsStruct ? (CSharpType)typeof(IPersistableModel<object>) : null;
+        }
+
+        public override string Name { get; }
+        public ModelTypeProvider Model { get; }
+        public JsonObjectSerialization? Json { get; }
+        public CSharpType IJsonModelTInterface { get; }
+        public CSharpType? IJsonModelObjectInterface { get; }
+        public CSharpType IPersistableModelTInterface { get; }
+        public CSharpType? IPersistableModelObjectInterface { get; }
+
+        protected override CSharpMethod[] BuildMethods()
+        {
+            return BuildSerializationMethods();
```

Comment:
nit: do we need the separate method given that this is a serialization provider? i.e. can we inline the method here?

Reply:
I think my original reasoning for this was that I wanted to clearly separate the code for serialization methods from any other methods we may need to generate, but since this class now pertains to all serialization methods I think it's unnecessary. I will move this inline.